news/2016-10-26-tip3-performance.asciidoc

```diff
@@ -0,0 +1,184 @@
+
+TIP: Most Weld-based runtimes allow to suppress implicit bean archives without `beans.xml`, i.e. require the `beans.xml` file in bean archives so that it's not necessary to scan all the parts of the application. See also link:/documentation/#4[FAQ].
+
+TIP: If it's not possible to use `annotated` mode, you can try a Weld-specific feature to conserve memory used: link:http://docs.jboss.org/weld/reference/latest/en-US/html/configure.html#veto-types-without-bean-defining-annotation[Veto types without bean defining annotation, window="_blank"].
+A similar solution is going to be standardized in CDI 2.0 (see also link:https://issues.jboss.org/browse/CDI-420[CDI-420, window="_blank"]).
+
+=== Extensions
+
+CDI portable extensions are essential integration points.
+And users love integrated technologies.
+However, the more extensions the more work must be done during bootstrap.
+For example, there is a `ProcessAnnotatedType` container lifecycle event.
+This event is fired:
+
+* for every type in an EXPLICIT bean archive,
+* for every session bean and every type with a bean defining annotation in an IMPLICIT bean archive.
+
+The delivery might be restricted by means of type arguments, e.g. for observer `void observeNumbers(@Observes ProcessAnnotatedType<Number> event)` an extension will be notified for every class assignable to `Number`.
+However, if you do `<T> void observe(@Observes ProcessAnnotatedType<T> event)` your extension will be notified for every class that is discovered.
+Now if we have an application with one EXPLICIT bean archive with 2000 classes, then a specific observer method on such extesion will be called 1000x.
+And if there three similar extensions, the CDI container will have to create 2000 events and notify 6000 observers.
+This should be ok unless the observer logic is overly complex.
+
+*CONCLUSION*: Pay attention to extensions used in your application.
+If you wonder what extensions are in service, use for example Weld Probe link:http://probe-weld.itos.redhat.com/weld-numberguess/weld-probe#/observers?bda=undefined&kind=EXTENSION[observers view, window="_blank"] or link:http://probe-weld.itos.redhat.com/weld-numberguess/weld-probe#/events[fired events view, window="_blank"] together with link:http://docs.jboss.org/weld/reference/latest/en-US/html/configure.html#config-dev-mode[`org.jboss.weld.probe.eventMonitor.containerLifecycleEvents` enabled, window="_blank"].
+
+TIP: If possible, don't use the extensions which observe all the annotated types from your application, i.e. which defined something like `void observe(@Observes ProcessAnnotatedType<?> event)`.
+Since CDI 1.1 `@WithAnnotations` should be used to restrict the set of types an extension is willing to process. link:http://probe-weld.itos.redhat.com/weld-numberguess/weld-probe#/observers?bda=undefined&kind=EXTENSION&observedType=javax.enterprise.inject.spi.ProcessAnnotatedType[Weld Probe, window="_blank"] warns you about these extensions.
```

Comment:
`extension is willing to` -> `extension is going to`
Unless the extension acquires its own will that is ;-)